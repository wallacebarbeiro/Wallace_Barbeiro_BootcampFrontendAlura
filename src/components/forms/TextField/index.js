import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import Text from '../../foundation/Text';

const InputWrapper = styled.div`
  margin-bottom: 17px;
`;

const Input = styled(Text)`
  width: 100%;
  border: 1px solid #444;
  background-color: transparent;
  padding: 12px 16px;
  outline: 0;
  border-radius: 0px;
  margin-bottom: 10px;
  color: #000;
  resize: none;
  

&::-webkit-input-placeholder  {
    color: #ccc;
}
    
&::-moz-placeholder {
    color: #ccc;
}
    
&:-ms-input-placeholder  {
    color: #ccc;
}
    
&:-moz-placeholder  {
    color: #ccc;
}
  
`;

Input.defaultProps = {
  tag: 'input',
  variant: 'paragraph1',
};

export default function TextField({
  placeholder,
  name,
  onChange,
  value,
  tag,
  type,
  error,
  isTouched,
  ...props
}) {
  const hasError = Boolean(error);
  const isFieldInvalid = hasError && isTouched;
  return (
    <InputWrapper>
      <Input
        as={tag}
        type={type}
        placeholder={placeholder}
        name={name}
        onChange={onChange}
        value={value}
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...props}
      />
      { isFieldInvalid && (
        <Text
          margin="0px"
          fontSize="12px"
          color="red"
          role="alert"
        >
          {error}
        </Text>
      )}
    </InputWrapper>
  );
}

TextField.propTypes = {
  placeholder: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
  type: PropTypes.string,
  tag: PropTypes.string,
  isTouched: PropTypes.bool,
  error: PropTypes.string,
};

TextField.defaultProps = {
  tag: 'input',
  type: 'text',
  error: '',
  isTouched: false,
};
