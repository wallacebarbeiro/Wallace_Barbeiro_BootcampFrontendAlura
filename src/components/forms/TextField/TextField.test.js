import React from 'react';
import user from '@testing-library/user-event';
import { render, screen } from '../../../infra/tests/testUtils';
import TextField from './index';

describe('<TextField />', () => {
  test('renders component', () => {
    render(
      <TextField
        placeholder="Nome"
        value="Mario"
        onChange={() => {}}
        name="nome"
      />,
    );

    // screen.debug();
    const textField = screen.getByPlaceholderText(/nome/i);

    expect(textField).toMatchSnapshot();
  });

  describe('when field is valid', () => {
    describe('and user is typing', () => {
      test('the value must be update', () => {
        const onChangeMock = jest.fn();
        render(
          <TextField
            placeholder="nome"
            value=""
            onChange={onChangeMock}
            name="nome"
            isTouched
          />,
        );

        const inputNome = screen.getByPlaceholderText(/nome/i);
        user.type(inputNome, 'Teste');
        expect(onChangeMock).toHaveBeenCalledTimes(5);
      });
    });
  });

  describe('when field is invalid', () => {
    test('display the respective error message', () => {
      render(
        <TextField
          placeholder="email"
          value="aa.com"
          onChange={() => {}}
          name="email"
          isTouched
          error="O campo e-mail é obrigatório"
        />,
      );
      const inputEmail = screen.getByPlaceholderText(/email/i);
      expect(inputEmail).toHaveValue('aa.com');

      const errorMessage = screen.getByRole('alert');
      expect(errorMessage).toHaveTextContent('O campo e-mail é obrigatório');

      expect(inputEmail).toMatchSnapshot();
    });
  });
});
