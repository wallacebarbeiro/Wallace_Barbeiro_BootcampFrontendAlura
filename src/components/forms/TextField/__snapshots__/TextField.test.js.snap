// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> renders component 1`] = `
.c0 {
  width: 100%;
  border: 1px solid #444;
  background-color: transparent;
  padding: 12px 16px;
  outline: 0;
  border-radius: 0px;
  margin-bottom: 10px;
  color: #000;
  resize: none;
}

.c0::-webkit-input-placeholder {
  color: #ccc;
}

.c0::-moz-placeholder {
  color: #ccc;
}

.c0:-ms-input-placeholder {
  color: #ccc;
}

.c0:-moz-placeholder {
  color: #ccc;
}

<input
  class="c0"
  name="nome"
  placeholder="Nome"
  type="text"
  value="Mario"
/>
`;

exports[`<TextField /> when field is invalid display the respective error message 1`] = `
.c0 {
  width: 100%;
  border: 1px solid #444;
  background-color: transparent;
  padding: 12px 16px;
  outline: 0;
  border-radius: 0px;
  margin-bottom: 10px;
  color: #000;
  resize: none;
}

.c0::-webkit-input-placeholder {
  color: #ccc;
}

.c0::-moz-placeholder {
  color: #ccc;
}

.c0:-ms-input-placeholder {
  color: #ccc;
}

.c0:-moz-placeholder {
  color: #ccc;
}

<input
  class="c0"
  name="email"
  placeholder="email"
  type="text"
  value="aa.com"
/>
`;
